04-cs-fundamentals/classroom/Binary.md

```diff
@@ -1,22 +1,277 @@
+
+  B<sub>16</sub>
+
+</details>
+
+
+
+## Levels of Abstraction
+
+Writing code in binary numbers or even hexidecimial numbers would be a tedious undertaking.  Therefore over time, programmers have added levels of abstraction to make programming easier.
+
+![Types of Languages](images/types-of-languages.png)
+
+<!-- Image source https://blog.malwarebytes.com/security-world/2012/09/so-you-want-to-be-a-malware-analyst/ -->
+
+The first level of abstraction is Assembly language.  In Assembly language each basic command the machine can understand is given a human readable name and commands have some basic formatting.  This is however very low level.  Below is an x86 hellow world assembly language program taken from [Loyola Marymount University](https://cs.lmu.edu/~ray/notes/x86assembly/).  Assembly code is run through an assembler which converts the text into machine code which can be executed.
+
+```asm
+
+          global    _start
+
+          section   .text
+_start:   mov       rax, 1                  ; system call for write
+          mov       rdi, 1                  ; file handle 1 is stdout
+          mov       rsi, message            ; address of string to output
+          mov       rdx, 13                 ; number of bytes
+          syscall                           ; invoke operating system to do the write
+          mov       rax, 60                 ; system call for exit
+          xor       rdi, rdi                ; exit code 0
+          syscall                           ; invoke operating system to exit
+
+          section   .data
+message:  db        "Hello, World", 10      ; note the newline at the end
+```
+
+From assembly programmers write the first high-level languages like C and Fortran.  These languages must be run through a _[compiler](https://en.wikipedia.org/wiki/Compiler)_ which is a program which translates the code into machine code.  A compiler differs from an assembler in that an assembler converts each command in a 1-to-1 manner.  Each assembly command cooresponds to a command the hardware understands.  In a compiled language however a single line may, like a loop, may represent multiple commands at the machine-code level.  
+
+Compare the following hello world C program with the above assembly program?  Which would you like to write?
+
+```c
+#include <stdio.h>
+int main()
+{
+   printf("Hello, World!");
+   return 0;
+}
+```
+
+Other languages like Ruby or Python are [interpreted languages](https://en.wikipedia.org/wiki/Interpreted_language) these languages are run by a special program called an _[interpreter](https://en.wikipedia.org/wiki/Interpreter_(computing))_ which reads the program and tells the computer what to do.  Because compiled languages are executed in machine-code form, they are often faster than interpreted languages, but an interpreted language does not need to be compiled for a specific type of computer.  An interpreted language can run on any computer with a valid interpreter.
```

Comment:
It might be worth mentioning that some languages can be interpreted _or_ compiled, since the split between interpreted and compiled languages can be a little fuzzy.  

It also could be worth mentioning a language like Java which is compiled into byte-code which is then interpreted.